diag_manager/fms_send_data_statfun.F90

```diff
@@ -0,0 +1,1139 @@
+         END IF MASK_VAR_IF
+      END ASSOCIATE
+
+!$OMP CRITICAL
+      IF ( .NOT.need_compute .AND. .NOT.reduced_k_range )&
+      & output_fields(out_num)%num_elements(sample) =&
+      & output_fields(out_num)%num_elements(sample) + (ie-is+1)*(je-js+1)*(ke-ks+1)
+      IF ( reduced_k_range ) &
+      & output_fields(out_num)%num_elements(sample) = output_fields(out_num)%num_elements(sample) +&
+      & (ie-is+1)*(je-js+1)*(ker-ksr+1)
+!$OMP END CRITICAL
+
+      succeded = .TRUE.
+      RETURN
+
+   END FUNCTION AVERAGE_THE_FIELD
+
+   FUNCTION SAMPLE_THE_FIELD (diag_field_id, field, out_num, &
+   & mask, sample, missvalue, missvalue_present, &
+   & l_start, l_end, idx_cfg, err_msg,  err_msg_local) result( succeded )
+      INTEGER, INTENT(in) :: diag_field_id
+      REAL, DIMENSION(:,:,:), INTENT(in) :: field
+      INTEGER, INTENT(in) :: out_num
+      LOGICAL, DIMENSION(:,:,:), INTENT(in), OPTIONAL :: mask
+      INTEGER, INTENT(in) :: sample
+      REAL, INTENT(in), OPTIONAL :: missvalue   !!TODO: optional?
+      LOGICAL, INTENT(in) :: missvalue_present
+      INTEGER, DIMENSION(3), INTENT(in)  :: l_start !< local start indices on 3 axes for regional output
+      INTEGER, DIMENSION(3), INTENT(in)  :: l_end !< local end indices on 3 axes for regional output
+      CHARACTER(len=*), INTENT(inout), OPTIONAL :: err_msg
+      CHARACTER(len=256), INTENT(inout) :: err_msg_local
+      TYPE(STATFUN_IDX_CFG_T), INTENT(in) :: idx_cfg
+      LOGICAL :: succeded
+
+      !!LOGICAL, ALLOCATABLE, DIMENSION(:,:,:) :: oor_mask
+      CHARACTER(len=128):: error_string
+
+      ! Power value for rms or pow(x) calculations
+      INTEGER :: pow_value, ksr, ker, is, js, ks, ie, je, ke, hi, hj, f1, f2, f3, f4
+      LOGICAL :: phys_window , need_compute , reduced_k_range
+
+      INTEGER :: i, j, k, i1, j1, k1
+
+      LOGICAL :: time_rms, time_max, time_min, time_sum
+#if defined(_OPENMP)
+      INTEGER :: omp_get_num_threads !< OMP function
+      INTEGER :: omp_get_level !< OMP function
+#endif
+
+      phys_window = output_fields(out_num)%phys_window
+      need_compute = output_fields(out_num)%need_compute
+      reduced_k_range = output_fields(out_num)%reduced_k_range
+
+      ! Is this output field the rms?
+      ! If so, then average is also .TRUE.
+      time_rms = output_fields(out_num)%time_rms
+      ! Looking for max and min value of this field over the sampling interval?
+      time_max = output_fields(out_num)%time_max
+      time_min = output_fields(out_num)%time_min
+      ! Sum output over time interval
+      time_sum = output_fields(out_num)%time_sum
+
+      ksr= l_start(3)
+      ker= l_end(3)
+      is = idx_cfg%is
+      js = idx_cfg%js
+      ks = idx_cfg%ks
+      ie = idx_cfg%ie
+      je = idx_cfg%je
+      ke = idx_cfg%ke
+      hi = idx_cfg%hi
+      hj = idx_cfg%hj
+      f1 = idx_cfg%f1
+      f2 = idx_cfg%f2
+      f3 = idx_cfg%f3
+      f4 = idx_cfg%f4
+
+      ASSOCIATE( OFB => output_fields(out_num)%buffer)
```

Comment:
This is old code and the new diag_object will need to be included here somehow.

Reply:
I am a little concerned about this `ofb` pointer creating a race condition.  How do you know that it isn't being re-associated by a different thread?

Reply:
Concerning the ofb association construct, you had suggested earlier to instead pass in the arrays as
arguments to the function.  I am going to try that approach next.